ftp/src/main/scala/akka.stream.contrib.ftp/FtpSource.scala

```diff
@@ -0,0 +1,151 @@
+import akka.stream.Attributes.{ InputBuffer, name }
+import akka.stream.contrib.ftp.FtpConnectionSettings.{ BasicFtpConnectionSettings, DefaultFtpConnectionSettings, DefaultFtpPort }
+import akka.stream.contrib.ftp.FtpCredentials.{ AnonFtpCredentials, NonAnonFtpCredentials }
+import akka.stream.scaladsl.Source
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, OutHandler }
+import org.apache.commons.net.ftp.{ FTPClient, FTPListParseEngine }
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+object FtpSource {
+
+  final val SourceName = "ftpSource"
+
+  def apply(): Source[FtpFile, Future[Long]] = apply(DefaultFtpConnectionSettings)
+
+  def apply(hostname: String): Source[FtpFile, Future[Long]] = apply(hostname, DefaultFtpPort)
+
+  def apply(hostname: String, port: Int): Source[FtpFile, Future[Long]] =
+    apply(BasicFtpConnectionSettings(InetAddress.getByName(hostname), port, AnonFtpCredentials))
+
+  def apply(hostname: String, username: String, password: String): Source[FtpFile, Future[Long]] =
+    apply(hostname, DefaultFtpPort, username, password)
+
+  def apply(hostname: String, port: Int, username: String, password: String): Source[FtpFile, Future[Long]] =
+    apply(BasicFtpConnectionSettings(InetAddress.getByName(hostname), port, NonAnonFtpCredentials(username, password)))
+
+  def apply(connectionSettings: FtpConnectionSettings): Source[FtpFile, Future[Long]] =
+    Source.fromGraph(new FtpSource(connectionSettings)).withAttributes(name(SourceName))
+
+}
+
+final class FtpSource private (connectionSettings: FtpConnectionSettings)
+  extends GraphStageWithMaterializedValue[SourceShape[FtpFile], Future[Long]] {
+  import FtpSource._
+
+  val matValue = Promise[Long]()
+
+  val shape = SourceShape(Outlet[FtpFile](s"$SourceName.out"))
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val InputBuffer(initialBuffer, maxBuffer) =
+      inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+
+    val logic = new GraphStageLogic(shape) {
+      import shape._
+
+      private var ftpClient: FTPClient = null
+      private var ftpListParseEngine: FTPListParseEngine = null
+      private var buffer: Vector[FtpFile] = Vector.empty
+      private var numFilesTotal: Long = 0L
+
+      override def preStart(): Unit = {
+        super.preStart()
+        try {
+          connect(
+            connectionSettings.host,
+            connectionSettings.port,
+            connectionSettings.credentials.username,
+            connectionSettings.credentials.password
+          )
+          ftpListParseEngine = ftpClient.initiateListParsing()
+          fillBuffer(initialBuffer)
+        } catch {
+          case NonFatal(t) =>
+            disconnect()
+            matFailure(t)
+            failStage(t)
+        }
+      }
+
+      override def postStop(): Unit = {
+        disconnect()
+        super.postStop()
```

Comment:
It's often good to complete the mat value promise from post stop to make sure that it's always completed.
Can be good to use `trySuccess` `tryFailure` when there are many code paths to completions.